op-program/host/prefetcher/prefetcher.go

```diff
@@ -253,6 +341,28 @@ func (p *Prefetcher) prefetch(ctx context.Context, hint string) error {
 			return err
 		}
 		return p.storeTransactions(txs)
+	case l2.HintL2Output:
+		if len(hintBytes) != 32 {
+			return fmt.Errorf("invalid L2 output hint: %x", hint)
+		}
+		hash := common.Hash(hintBytes)
+		output, err := p.l2Fetcher.OutputByRoot(ctx, hash)
+		if err != nil {
+			return fmt.Errorf("failed to fetch L2 output root %s: %w", hash, err)
+		}
+		return p.kvStore.Put(preimage.Keccak256Key(hash).PreimageKey(), output.Marshal())
+	}
+
+	// some L2 state data can be fetched in bulk from block execution witnesses instead of direction from the MPT
+	// if we have a bulk hint, we should use it instead of the last hint (will fallback to last hint after bulk hint is cleared and request is retried)
+	if p.lastBulkHint != "" && p.experimentalSourceEnabled {
+		bulkHint := p.lastBulkHint
+		p.lastBulkHint = ""
+		return p.bulkPrefetch(ctx, bulkHint)
+	}
+
+	switch hintType {
```

Comment:
I agree it's harder to read. I can extract the prefetchers for state/code to a separate function and handle the bulk hint there, then I'll just keep the cases in the same switch case which will call the new function.

I don't think we should always execute the bulk hint when it's received since the state nodes could be in the cache or could have been fetched by another execution witness/account proof hint already.

Reply:
Let me know what you think of this: edb17fe

I think this prevents these hints from acting like commands.